### ISSUE TITLE
Testing BTC-RPC-Proxy to simulate an unpruned node

### ISSUE THREAD

**Issue by openoms** [cryptoadvance/specter-desktop#722]
From: https://github.com/cryptoadvance/specter-desktop/issues/298#issuecomment-724137049
>This code https://github.com/Start9Labs/btc-rpc-proxy-wrapper (though in another language) is able to fetch pruned blocks from other peers.
Based on: https://github.com/Kixunil/btc-rpc-proxy (written in Rust)

This could help Specter to run a with pruned node and maybe would make the use of external block explorers unnecessary (see https://github.com/cryptoadvance/specter-desktop/issues/718)

Quick installation steps as I did on Debian Linux: https://gist.github.com/openoms/2a94a8ad756bf93c8cd57d68958450a6#file-btc-rpc-proxy-md

Need to allow all RPC commands used. Got these when searched for the self.rpc.commands in specter-desktop: https://gist.github.com/openoms/2a94a8ad756bf93c8cd57d68958450a6#configure , but likely missing something still.

When running Specter (`python3 -m cryptoadvance.specter server`) the detection of the bitcoinRPC on localhost (BTC-RPC-proxy) works and connection is tested successfully but immediately runs into these errors:
https://gist.github.com/openoms/2a94a8ad756bf93c8cd57d68958450a6#file-specter-desktop-log

Same time errors in BTC-RPC-Proxy are showing the bad requests:
https://gist.github.com/openoms/2a94a8ad756bf93c8cd57d68958450a6#file-btc_rpc_proxy-log

Either using BTC-RPC-Proxy is really close and I am just missing something or the next step could be to request @Kixunil for the feature if we figure it out what is exactly needed.

**Comment by Kixunil** [comment]
Hi, happy you're interested!

> Quick installation steps as I did on Debian Linux

You don't need to install Rust by piping commands to `sh` on Debian 10. Just `apt install cargo`. You need to use `--features=compat` during `cargo` `run`/`build`/`install` but that should be fine.

I'm now looking into re-packaging it in [my Debian repository](https://deb.ln-ask.me), so you will have even easier time installing. :)

Also your instructions describe cloning `master`, which **does not yet** contain block fetching. Not sure if mistake in the instructions or you actually attempted that. The new version is here: https://github.com/Start9Labs/btc-rpc-proxy/tree/feature/refactor

Regarding configuration, daemon cookie authentication is supported and [recommended](https://github.com/Kixunil/security_writings/blob/master/cookie_files.md). I plan to implement it for users as well eventually. Also you seem to be missing `fetch_blocks = true` in `[user.specter]` section. (Should be unrelated to error messages you posted.)

> but immediately runs into these errors:

Would you be willing to try again and use [Wireshark](https://www.wireshark.org/) to dump the exact request? (right click on one of the request packets -> follow TCP stream) (feel free to censor private information)
There's a possibility it got fixed during refactor but I'd like to understand what I was doing wrong so that I can prevent it in the future.
Then if you actually tried the old version, try again with the new version.

> and I am just missing something

It's still an alpha feature. :) But it's intended to be entirely transparent so that the application doesn't even know it's not a real bitcoind. Therefore if there's any problem, definitely ping us.

**Comment by openoms** [comment]
Thank you will test and update the instructions.

did now:
```
git clone https://github.com/Start9Labs/btc-rpc-proxy.git
cd /home/brp/btc-rpc-proxy
git checkout feature/refactor
```

> You don't need to install Rust by piping commands to sh on Debian 10. Just apt install cargo. You need to use --features=compat during cargo run/build/install but that should be fine.

`$ cargo build --features=compat --release`
```
error: Package `btc-rpc-proxy v0.2.3 (/home/brp/btc-rpc-proxy)` does not have these features: `compat`
```

and I get this with:
`~/btc-rpc-proxy$ cargo build --release`
```
   Compiling btc-rpc-proxy v0.2.3 (/home/brp/btc-rpc-proxy)
error[E0658]: use of unstable library feature 'str_strip': newly added
   --> src/rpc_methods.rs:207:22
    |
207 |                 host.strip_suffix(".onion")
    |                      ^^^^^^^^^^^^
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/67302

error[E0658]: use of unstable library feature 'str_strip': newly added
  --> src/users.rs:27:31
   |
27 |         let auth = header_str.strip_prefix("Basic ")?;
   |                               ^^^^^^^^^^^^
   |
   = note: for more information, see https://github.com/rust-lang/rust/issues/67302

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
error: could not compile `btc-rpc-proxy`.

To learn more, run the command again with --verbose.
```

**Comment by Kixunil** [comment]
`git clone -b feature/refactor` or `git checkout feature/refactor` ;)

**Comment by openoms** [comment]
> `compat` was renamed to `old_rust` a few hours ago and I didn't notice it.

ok so instead of `cargo build --release` using now:
`cargo build --features=old_rust` 
but only made one step further:
```
   Compiling btc-rpc-proxy v0.2.3 (/home/brp/btc-rpc-proxy)
   Compiling bitcoin v0.25.2
error: reached the type-length limit while instantiating `std::pin::Pin::<&mut std::future...}[1]::poll[0]::{{closure}}[0])]>`
    |
    = note: consider adding a `#![type_length_limit="1717928"]` attribute to your crate

error: aborting due to previous error

error: could not compile `btc-rpc-proxy`.

To learn more, run the command again with --verbose.
```

**Comment by Kixunil** [comment]
Ah, yeah, I reported it and it seems they forgot to add it. The suggestion from the compiler works, so for the time being you can manually add  `#![type_length_limit="1717928"]` line at the top of `src/main.rs`

**Comment by openoms** [comment]
now on connecting from Specter clicking test:
```
[2020-11-30 14:09:13,818] ERROR in specter: Caught an RpcError while test_rpc: Server responded with error code 500: {"id":null,"error":{"code":-1,"message":"error trying to connect: tcp connect error: Connection refused (os error 111)"},"result":null}
[2020-11-30 14:09:13,818] ERROR in specter: 500
```

BTC-RPC-Proxy shows no errors:
```
$ cargo run --features=old_rust -- --conf=/home/brp/btc-rpc-proxy/btc_rpc_proxy.toml
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/btc_rpc_proxy --conf=/home/brp/btc-rpc-proxy/btc_rpc_proxy.toml`
Nov 30 14:08:55.680 DEBG specter called getnetworkinfo: FORWARDED
Nov 30 14:09:03.898 DEBG specter called getnetworkinfo: FORWARDED
Nov 30 14:09:13.816 DEBG specter called getnetworkinfo: FORWARDED
```

**Comment by openoms** [comment]
It works now:
![image](https://user-images.githubusercontent.com/43343391/100622546-e3dbda00-3318-11eb-99b5-9c7cb4cad19e.png)

Can see a wall of RPC calls passing through the Proxy:
![image](https://user-images.githubusercontent.com/43343391/100622701-11288800-3319-11eb-9fb5-67eb0774c560.png)

no errors in the Specter server and no popup about the pruned node either!

Updating instructions and will explore more.

**Comment by openoms** [comment]
Next problem is that Specter still perceives as it is a pruned node:
![image](https://user-images.githubusercontent.com/43343391/100623780-86e12380-331a-11eb-9bf3-7fc607456ca9.png)
Specter server:
```
ERROR in controller: Request error: Can't rescan beyond pruned data. Use RPC call getblockchaininfo to determine your pruned height. while rescanblockchain
```
No errors in the Proxy.

**Comment by Kixunil** [comment]
Nice!!! I still want to test it a bit more before merging and signing as the official version. Perhaps doing a second pass of review as well. If you know of other projects that need non-pruned node, then you can help push this forward by trying out the proxy with them as well and reporting your findings.

I also want to add a new configuration option for log level later so that production users are not spammed with those messages. The logs are quite useful for the time being.

**Comment by Kixunil** [comment]
Weird, GitHub showed me your latest message only after I commented...

Looks like specter attempted wallet a rescan and that seems to be unimplemented in `btc-rpc-proxy`.
@dr-bonez are you aware of this? I think we should support it somehow, although it'll likely be heavy - will need to re-download whole chain. Perhaps caching block filters could be a reasonable trade-off.

**Comment by dr-bonez** [comment]
In order to enable the "unpruned" feature, you need to add `fetch_blocks = true` to the config for your user

**Comment by Kixunil** [comment]
Thinking about it again, I think adding `electrs` support as I suggested in the PR could help with this too. Yeah, `electrs` has 68G on my machine but that's still less than 365G of chain. And rescan would become *much* quicker. What do you think?

**Comment by openoms** [comment]
now tried to import my backed up wallets while connected to BTC-RPC-Proxy backed by a real unpruned bitcoind node.
It went on to rescan indefinitely repeating this in the Specter logs:
```
INFO in wallet: Didn't get transactions on address xxxxxx. Refilling keypool.
WARNING in wallet_manager: Couldn't find wallet xxxxx in core's wallets.Silently ignored!
WARNING in wallet_manager: Couldn't find wallet xxxxx in core's wallets.Silently ignored!
...
WARNING in wallet_manager: Couldn't find wallet xxxxx in core's wallets.Silently ignored!
WARNING in wallet_manager: Couldn't find wallet xxxxxr_3 in core's wallets.Silently ignored!
INFO in wallet: Didn't get transactions on address xxxxxxxx. Refilling keypool.
```
The scanning was triggered in bitcoind, but as it seems the wallets couldn't be saved while going through the Proxy.
There were no errors in the Proxy logs.

Specter loads and unloads wallets to bitcoind and that must be not playing well together with BTC-RPC-Proxy.

**Comment by Kixunil** [comment]
Proxy should have no effects on this behavior. It looks like you forgot to restore the wallet files or something.
If the content of the bitcoind directory looks OK (contains the wallet xxxx), then please try same setup, just without proxy - set your node directly in Specter.

**Comment by openoms** [comment]
You are right, it was a problem with restoring with a used bitcoind wallet. I've recreated the bitcoind wallet.dat and emptied the wallets directory from .specter and the errors are gone.
Works through the BTC-RPC-proxy too.

I am inclined to close this issue since it is not (yet) possible to do rescan on a pruned node without redownloading the full blockchain or plugging an electrs instance to either Specter itself or to BTC-RPC-Proxy.

Looking forward to the further developments on BTC-RPC-Proxy and it was well proven that it is capable of setting the fine-grained bitcoin RPC permissions as it was intended to.

**Comment by Kixunil** [comment]
I'm not sure if this is even solvable problem at all. If you don't have any kind of index or wallet metadata, then downloading the whole chain (in theory does not have to be stored) is the only trustless way to recover a wallet.

But if you're still interested in some of the other tradeoffs available, I'd love to hear your opinion which of them sounds most reasonable to you so that maybe I can look into it one day.

Options that seem to be realistic:

* Download whole chain (without storing - consumes bandwidth, not storage) and do trivial rescan
* Try to fetch block filters from peers and download relevant blocks
* Try to cache block filters during the lifetime of the node and use them. The question is what should happen if there are missing filters.
* Use `electrs` to get the relevant transactions/blocks

**Comment by openoms** [comment]
I think 
> Try to fetch block filters from peers and download relevant blocks

would be the most desirable option given there would be more nodes serving them.
Also this option was I heard @stepansnigirev talk about on this recent  podcast: https://stephanlivera.com/episode/231/.

**Comment by Kixunil** [comment]
Yeah, that sounds pretty good to me too. We might also check if bitcoind can keep them despite pruning and set it to do so. But AFAIK only Knots can send filters over RPC.

**Comment by dr-bonez** [comment]
>Try to fetch block filters from peers and download relevant blocks

just a note, this is a privacy leak. Ideally it should be configurable as to whether you download the whole chain, or only blocks that you need.

**Comment by Kixunil** [comment]
Not extremely big privacy leak, I believe. It should be hard to correlate transactions from whole blocks (which even have false-positives) and each block (candidate) could be downloaded over separate Tor connection.

Having it configurable for paranoid users would be nice of course.

**Comment by dr-bonez** [comment]
Simulating `blockfilterindex` (providing the `getblockfilter` rpc) for a pruned node would be a huge win, since I bet it will be a long time before such a thing ends up in core, however it would change the role of proxy from being just a light proxy, to something that needs to sync. If you want to go that route though I'd be happy to help.

**Comment by Kixunil** [comment]
I'm definitely not opposed to the idea of having the proxy a bit heavier if it is configurable (opt-in) and the code is properly organized. Especially if it means Bitcoin users no longer have to store 300G of chain data to access very useful Bitcoin features. I don't think I can find the time to work on it myself in the following month or two but I could review/merge a PR if you're motivated to do it sooner.

**Comment by dr-bonez** [comment]
FYI, @chrisguida is working on this. It requires giving proxy control over pruning though, to do it efficiently, so that it can ensure undo files are not deleted before they are used.

**Comment by Kixunil** [comment]
That's great news! I'm not very happy with controlling pruning over RPC as it'll either break my setup or I'll have to do some serious redesign. Would be nice if fallback was possible.

But I do realize the benefits so I **do** intend to think about redesigning. It'll just take some time to do it.

**Comment by chrisguida** [comment]
@Kixunil FYI electrs [requires](https://github.com/romanz/electrs/blob/master/doc/usage.md#bitcoind-configuration) a full archival bitcoin node.

**Comment by Kixunil** [comment]
@chrisguida yes I know, [I wrote that part of the docs](https://github.com/romanz/electrs/blame/master/doc/usage.md#L160) :) It should be compatible with RPC proxy faking non-pruning though as long as `jsonrpc_import` is used. Didn't test yet but would be highly surprised if it didn't work.

**Comment by dr-bonez** [comment]
@Kixunil the alternative is to maintain a second UTXO set, separate from core in order to know the scripts associated with spends. It's much slower and takes a lot more storage.

**Comment by Kixunil** [comment]
@dr-bonez ah, didn't realize those scripts are needed to build block filters. Would appreciate if it was possible to turn this feature off which would also stop control of pruning. At least for some time.

What I'm more concerned about is that if a user has already installed pruned version and wants to start using btc-rpc-proxy later, it'd fail. So some kind of fallback even if slow and temporarily using more storage would be useful. I can understand if this is not priority, just something to keep in mind and perhaps design the code to not be hard to add.

**Comment by dr-bonez** [comment]
oh for sure. All of these kinds of features should be opt in. But they won't get block filters if they don't turn it on.

**Comment by chrisguida** [comment]
@Kixunil 

Ah, my apologies, I hadn't noticed that 😅

**Comment by Kixunil** [comment]
Waiting for 0.22 is not too bad.

@chrisguida no problem, no offense taken. Just had some laugh. :)
